Dev

Jasonzhouquan opened the pull request:
I figured that I would just close the previous one....

Jasonzhouquan committed:
Daily update with comments

* added dryrun in Tomography for "summarize_plan"

* added comments for more TODO items

Jasonzhouquan committed:
daily updates

* Energy foil is a wheel device that get s used once per run (probably)
* FS would probably be mounted on at least one motor.
* We need to reconsider the use of the fast shutter since the FS continuous used during the scan and are triggered differently according to experiment.

Jasonzhouquan committed:
Changed to PointGrey Detector

Jasonzhouquan committed:
Added comments on Calibration and alignment

Jasonzhouquan committed:
Updated beamline equipments

* Updated with two slits (upstream/downstream)

* updated some attenuator info

* added focus lenses info from 1-ID, not sure how many lenses we will have and what kind of setup yet.

Jasonzhouquan committed:
updated some of the atten section and focuslenses

* The attenuator is more frequently used and not changed in each scan (typically).

* Consider provide user access to find the optimum attenuation and then read the correct value in config yml file.

* the slits are similar to the attenuators but less frequently changed

* We will have four focusing lenses, each siiting on top of at least 5 motor. x,y,z, and tiltx, tilty/rot, tiltz.  Let's discuss more on this...

Jasonzhouquan committed:
comments on improvements

Jasonzhouquan committed:
Added some 6BM code

Jasonzhouquan committed:
added comments to take care of the beamline property before experiment.

Jasonzhouquan committed:
updated how the ymls is structured

Jasonzhouquan committed:
update with chen

* improved attenuator control

* comments added for future improvements

Jasonzhouquan committed:
minor fix and comments

Jasonzhouquan committed:
comment

Jasonzhouquan committed:
formated as discussed

Jasonzhouquan committed:
updated beamline.py as discussed with Chem

* finished attenuator structure

* added four lenses with 6 motors each

* changed slit motors to more meaningful names

Jasonzhouquan committed:
topo fix

Jasonzhouquan committed:
added slits, atten, and focuslenses to the Beam class

Jasonzhouquan committed:
Update tomo_scan_template.yml

added beamsize control in the template

Jasonzhouquan committed:
Update beamline.py

* updated h_size and v_size control for the two slits wiht beamline 1 PVs.

* just as place holders

Jasonzhouquan committed:
Update beamline.py

typo fix

Jasonzhouquan committed:
Update experiment.py

1. added tomobeam=Beam()

2. set up simulated motors for Beam control in 'debug' mode

3. costmetic changes overall

4. added step 0.5 to set up slits (currently with the h_size control, not sure if this is an EpicsMotor);  to set attenuation;  and to check energy ( still not clear the exact implementation)

Jasonzhouquan committed:
Update experiment.py

minor fix

Jasonzhouquan committed:
Update motors.py

* updated the motor stack configuration

* changed the motor names for easy access and clarification.

* added the Kohzu tilt motors above the Aero.

* updated other items accordingly

Jasonzhouquan committed:
Update beamline.py

changed 'break' to 'raise ValueError' for attenuation out of range

Jasonzhouquan committed:
Update beamline.py

cosmetic changes

Jasonzhouquan committed:
Update motors.py

cosmetic changes

Jasonzhouquan committed:
Update tomo_scan_template.yml

updated motor names

Jasonzhouquan committed:
Update experiment.py

*changed name preci to aeri

* added getbeam part in NearField class

Jasonzhouquan committed:
Update experiment.py

updated legacy 6BM sample motor names with StateAero motor names.

Jasonzhouquan committed:
Update experiment.py

added get detector part in NearField

Jasonzhouquan committed:
Update experiment.py

added white field collection

need to confirm with the current 1-id routine

Jasonzhouquan committed:
update file name and the nf part

updated nf config for scanning volume set up

Jasonzhouquan committed:
Update scan_template.yml

no white field for NF

Jasonzhouquan committed:
Update beamline.py

added a comment based on conversation with Peter

focuslenses are typically only allowed to be changed in the y direction

Jasonzhouquan committed:
Update experiment.py

added flyscan and darkfield collection

Jasonzhouquan committed:
update with the two z nf positions

Jasonzhouquan committed:
updated NF control

* currently output the same file containing images at both Z posiitons

* not sure if the function calling in the for loop will work

* deleted white field and dark field for nf.

Jasonzhouquan committed:
Update scan_template.yml

recinfigured FF part

Jasonzhouquan committed:
Update experiment.py

* updated with feature to repeat the single layer scan for the current layer by setting ky_step to 0.

* added the FF part
   > Currently, the ff part will perform multiple layers
   > For each layer, dark field images will be collected before and after the layer scan
   > Dexela needs to be implemented.
   > Calibration need to be clarified.

Jasonzhouquan committed:
Update detectors.py

Fixed the structure from a previous (wrong) imput

Added my best wishes of not going with Retiga

Jasonzhouquan committed:
Update experiment.py

requesting new features

Jasonzhouquan committed:
Update experiment.py

* deleted interactive feature per discussion with Chen

Jasonzhouquan committed:
Update experiment.py

minor fixes

Jasonzhouquan committed:
Update experiment.py

code review and typo fix

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+tomo:
+  ## Beamline optics
+  energyfoil:   "Re"      #   Ho, Pb, Re, Tb, Air, Au, Yb, Bi, Pt, Ir, Hf, Tm, Ta
+                          #   probably just need to do once???
```
How about naming these energy foils following this type of convention `$Element_$Energy`?

Jasonzhouquan replied:
Per previous conversation, I will remove the beam optics section from the yml file

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+                          #   probably just need to do once???
+  attenuation:  "a number"         # Place hodler. This doesn't seem to be pre-defined for the entire run
+  FS_control:   "sweep"   # sweep, detpulse, detexposure etc. used this to determine FS control
+  beamsize_h:   0.5       # horizontal beam size
```
What are the __units__ for the beam size?  _Pixels_ is probably the easiest way to go, but it does require us to provide unit conversion somewhere else for the users.

Jasonzhouquan replied:
The beam size is from the slit motor set up, which is in mm (typically). I don't think we should use pixels, because that would depend on the actual detector used.  The physical dimensions of the slits would make more sense as this relates to the sampling size.

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+  beamsize_v:   0.5       # vertical beam size
+  ## Scan parameters
+  type:    fly           # [step|fly]
+  n_white: 5             # num of white field before and after tomo
```
What if we always just take 10/20/30 white&dark field images instead of ask the users to specify?  In other words, remove this option from config to further simplify the input. I don't think the actual number of bg is something so critical to the users that they have to specify.

Jasonzhouquan replied:
This is a good point. I think we need to minimize the user input to avoid confusion (and typos) here and there. 
I will keep them there for now but make a comment so that we remove all the non-important/constant inputs later.

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+  # below are for fly_scan only
+  ROT_STAGE_FAST_SPEED:       1   # degree/second,
+  accl:                       3   # second,
+  readout_time:  0.001    # detector readout time
```
Since the `readout_time` is an intrinsic property of each camera, how about we use the camera names (if we need to support multiple cameras) or just directly hard-code them in?

Jasonzhouquan replied:
Let wait until we get the camera and confirm this.  I'm not entirely sure if this number is a constant or if it changes over time.
Either way, I'm adding a note for this.

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+  ROT_STAGE_FAST_SPEED:       1   # degree/second,
+  accl:                       3   # second,
+  readout_time:  0.001    # detector readout time
+
```
the questions in the Tomo section also apply here

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+
+output:  # we should include the output in each experiment (i.e. tomo, NF, FF)       
+#  filepath:    '/home/beams/S6BM/user_data/2019-2/startup_jun19/tomo/'    # use testing location
+  filepath:    '/dev/shm/tmp/'    # use testing location
```
Since we are considering a quick local NAS for data caching, perhaps a simple directory/project name should be used instead of full path.

Jasonzhouquan replied:
This I'm not quite sure.... but I think it is simple enough right now for the users to change things with a very small amount of input.

KedoKudo commented on seisidd/config/scan_template.yml:
```diff
@@ -0,0 +1,103 @@
+output:  # we should include the output in each experiment (i.e. tomo, NF, FF)       
+#  filepath:    '/home/beams/S6BM/user_data/2019-2/startup_jun19/tomo/'    # use testing location
+  filepath:    '/dev/shm/tmp/'    # use testing location
+  fileprefix:  'ttt'       # specify file name
```
Following the trajectory of the previous comment, we might be able to extract the name directly from the proposal.  However, I am not sure whether the users would like this simplified approach or not.

Jasonzhouquan replied:
At least from personal experience, a custom file path set up is useful when the user want to take test images, extra dark/sample patterns is a different directory etc.

Jasonzhouquan replied:
But this may work for the scans only. I will add a comment and see how things going once we have the setup.

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -9,48 +9,281 @@
 
-class Slits(MotorBundle):
-    """Control the four blades that form the beam"""
+class SlitUpstream(MotorBundle):
```
seems like we are going with the Singleton approach, which should be fine as along as we do not have too many similar devices.

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -9,48 +9,281 @@
-class Attenuator(MotorBundle):
+### Per Peter, only motion in the y direction is allowed to avoid accidents
+### consider comment out all the other PV accesses.
+class FocusLens1(MotorBundle):
```
These FocusLensX classes are very similar to each other.  So we might be better off combining them into one.

Jasonzhouquan replied:
I will comment out these focus lenses since they shouldn't be accessed by users. 
I will leave out the controls in the y direction for switching to focus beams.

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -9,48 +9,281 @@
     def __init__(self):
-        self.slits = Slits(name='slits')
-        self.atten = Attenuator()         # adjustment needed
+        self.s1         = SlitUpstream()
```
it seems like the two slits, four focus lens are only used inside beam.  What if we make the slits and focus lenses embedded class instead, hiding them from accidental/unwanted access.

Jasonzhouquan replied:
I will remove the beam line control out of the RE plan.
The optics adjustments should be done in the interactive session inside Jupyter

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -9,48 +9,281 @@
     @property
     def status(self):
         """Return full beam status"""
+        #   Some info to print here:
```
using __repr__ as the status report might work better in a notebook env.

Jasonzhouquan replied:
Will replace all the status(self)

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -9,48 +9,281 @@
     #   Requir acutal implementation once the nature of the device is known
+    #   The following info is from 1ID
+    #   Motor 55 in B Crate: (5Ph OM motor, Res=-0.360, Speed=60)
+    #   Pos {Element} Thck K-edge, magic#, [range], I_SCU1, Harm, HEM(keV)
```
The information here are too important to stored as comments.  How about a wiki page or a dedicated documentation (PDF)?  At least, we should provide something like a man page to allow a quick search of these information.

Jasonzhouquan replied:
I agree, will put those up when things are finalized.

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -9,48 +9,281 @@
     pass
 
 
+#   May not need this
```
Then we should probably just comment it out for now.

Jasonzhouquan replied:
This is part of the energy calibration, I will keep this as a place holder

KedoKudo commented on seisidd/devices/beamline.py:
```diff
@@ -59,8 +292,9 @@ class Counter:
     """Fast shutter for taking dark field images"""
     # NOTE:
-    #  apstoos provides shutter class for the main/slow beamline shutter,
+    #  apstools provides shutter class for the main/slow beamline shutter,
```
If similar shutters are also used at other beamlines, it would be the best for @prjemian to provide support instead of individual beamilnes.

Jasonzhouquan replied:
Sure, I will check to see if the FS are a common device across the beamlines.

KedoKudo commented on seisidd/devices/detectors.py:
```diff
@@ -6,43 +6,66 @@
 NOTE:
 * Retiga area detector is not readily available in Ophyd
 * Retiga area detector camera needs to be either setup here or PR to ophyd 
+* Changed to PointGrey Detectors in this example
```
Perhaps we should make a request in Ophyd to ask for support of Retiga detector.

Jasonzhouquan replied:
Currently we are closer to the FLIR/PG option, but this won't hurt.

KedoKudo commented on seisidd/devices/motors.py:
```diff
@@ -9,66 +9,88 @@
         """return full pv list and corresponding values"""
         # TODO:
-        #   once acutal PVs are known, the implementation should goes here
+        #   once acutal PVs are known, the implementation should go here
```
use __repr__ instead of status()?  Or perhaps both?

Jasonzhouquan replied:
Let me investigate what this represents does...

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -108,8 +120,83 @@ def calibration(self):
+            #   This is a place holder for maybe additional control of the beam
+            #   in the debug mode.  Let's discuss what should be exposed here.
+            #   Or, we just use some simulated motors for the beam here as below?
+            from ophyd import sim
```
The current simulated devices from Ophyd is not very useful for us.  We are better off using simulated devices from Epics.

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -321,10 +425,46 @@ def tomo_scan(self, cfg):
+        #   epics_put("1ide1:Kohzu_E_dnHsize.VAL", (($1)+0.1), 10) ##
+        #   epics_put("1ide1:Kohzu_E_upVsize.VAL", ($2), 10) ## VERT SIZE
+        #   epics_put("1ide1:Kohzu_E_dnVsize.VAL", ($2)+0.1, 10) ##
+        _beam_h_size    =   cfg['tomo']['beamsize_h']
```
adjust the beam size during the Tomo scan?  Is there any particular good reason for doing this?

Jasonzhouquan replied:
Will remove this.

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -321,10 +425,46 @@ def tomo_scan(self, cfg):
+        # check energy
+        # need to be clear what we want to do here
+        _energy_foil = cfg['tomo']['energyfoil']
+        yield from bps.mv(beam.foil, _energy_foil)      # need to complete this part in beamline.py
```
It seems to me that the beamsize and beam energy should be pre-configured in the notebook.  Asking RE to handle this part feels dangerous to me.

Jasonzhouquan replied:
Agreed

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -401,16 +544,845 @@ def scan_closure():
+        self.fly_control    = NearField.get_flycontrol(self._mode)
+        self.nf_det         = NearField.get_detector(self._mode)
+        self.nf_beam        = NearField.get_nfbeam(self._mode)
+        # TODO:
```
Once everything is properly calibrated, we can rely on automated initialization during runtime.  However, I believe manual adjustment is unavoidable, and probably the best way to go for a long time.

Jasonzhouquan replied:
Let's wait for actual testing on this...

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -401,16 +544,845 @@ def scan_closure():
+            #   This is a place holder for maybe additional control of the beam
+            #   in the debug mode.  Let's discuss what should be exposed here.
+            #   Or, we just use some simulated motors for the beam here as below?
+            from ophyd import sim
```
similar issue with the simulated devices

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -401,16 +544,845 @@ def scan_closure():
+            # check the following page for important information
+            # https://github.com/BCDA-APS/use_bluesky/blob/master/notebooks/sandbox/images_darks_flats.ipynb
+            #
+            epics.caput("PV_DET:cam1:FrameType.ZRST", "/exchange/data_white_pre")
```
for different experiment, we might need to set different layout (folder names) through the same PVs.  This is just something that we have to keep in mind.

Jasonzhouquan replied:
Sure, I'm much less familiar with this.  Let me know what you need to change if that happens.

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -401,16 +544,845 @@ def scan_closure():
+    @bpp.run_decorator()
+    def scan_singlelayer(self, cfg_nf, _layer_number):
+        # TODO:
+        #   Somewhere we need to check the light status, or, add a suspender?
```
yes, but we cannot instantiate anything until the system is in-place (devices and soft-IOCs).

Jasonzhouquan replied:
Sure, this is a reminder for myself

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -401,16 +544,845 @@ def scan_closure():
+        n_layers   = cfg['nf']['volume']['n_layers']
+        ky_start   = cfg['nf']['volume']['ky_start']
+        ky_step    = cfg['nf']['volume']['ky_step']
+        if ky_step == 0:
```
for repeating scans, there is already one plan stub available.  If we want to do multiple detector distances, the movement of the nf-det should be included in the scan plan.

Jasonzhouquan replied:
can you share the repeating scan stub from Bluesky?  I'll check if I can easily apply that instead of this homemade approach...

Jasonzhouquan replied:
The repeat scan from Bluesky seems different from what I wanted. will keep my original script for now

KedoKudo commented on seisidd/experiment.py:
```diff
@@ -401,16 +544,845 @@ def scan_closure():
+            #   This is a place holder for maybe additional control of the beam
+            #   in the debug mode.  Let's discuss what should be exposed here.
+            #   Or, we just use some simulated motors for the beam here as below?
+            from ophyd import sim
```
similar issues with simulated devices.

KedoKudo reviewed (approved):
There are a lot of places need to be adjusted.  But as a starting point, this is solid enough.

Jasonzhouquan committed:
Update scan_template.yml

* edited and added comments according to suggestions

Jasonzhouquan committed:
Update scan_template.yml

added comment

Jasonzhouquan committed:
Update scan_template.yml

comment on filepath auto fill

Jasonzhouquan committed:
Update beamline.py

commented out all the motors except for the y motors.

consider direct define only the y control of the lenses.

Jasonzhouquan committed:
Update beamline.py

Jasonzhouquan committed:
Update experiment.py

removed beamline optics setting inside RE,

instead, RE will read the current optics setting and report it.
